components/zcash_note_encryption/src/lib.rs

```diff
@@ -1,8 +1,474 @@
+/// `PaymentAddress` to which the note was sent.
+///
+/// Implements section 4.17.2 of the Zcash Protocol Specification.
+pub fn try_note_decryption<D: Domain, Output: ShieldedOutput<D>>(
+    domain: &D,
+    ivk: &D::IncomingViewingKey,
+    output: &Output,
+) -> Option<(D::Note, D::Recipient, D::Memo)> {
+    assert_eq!(output.enc_ciphertext().len(), ENC_CIPHERTEXT_SIZE);
+
+    let shared_secret = D::ka_agree_dec(ivk, output.epk());
+    let key = D::kdf(shared_secret, output.epk());
+
+    let mut plaintext = [0; ENC_CIPHERTEXT_SIZE];
+    assert_eq!(
+        ChachaPolyIetf::aead_cipher()
+            .open_to(
+                &mut plaintext,
+                output.enc_ciphertext(),
+                &[],
+                key.as_ref(),
+                &[0u8; 12]
+            )
+            .ok()?,
+        NOTE_PLAINTEXT_SIZE
+    );
+
+    let (note, to) = parse_note_plaintext_without_memo_ivk(
+        domain,
+        ivk,
+        output.epk(),
+        &output.cmstar(),
+        &plaintext,
+    )?;
+    let memo = domain.extract_memo(&plaintext);
+
+    Some((note, to, memo))
+}
+
+fn parse_note_plaintext_without_memo_ivk<D: Domain>(
+    domain: &D,
+    ivk: &D::IncomingViewingKey,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+    plaintext: &[u8],
+) -> Option<(D::Note, D::Recipient)> {
+    let (note, to) = domain.parse_note_plaintext_without_memo_ivk(ivk, &plaintext)?;
+
+    if let NoteValidity::Valid = check_note_validity::<D>(&note, epk, cmstar) {
+        Some((note, to))
+    } else {
+        None
+    }
+}
+
+fn check_note_validity<D: Domain>(
+    note: &D::Note,
+    epk: &D::EphemeralPublicKey,
+    cmstar: &D::ExtractedCommitment,
+) -> NoteValidity {
+    if &D::extract_note_commitment(&note) != cmstar {
+        // Published commitment doesn't match calculated commitment
+        NoteValidity::Invalid
+    } else {
+        let epk_bytes = D::epk_bytes(epk);
+        D::check_epk_bytes(&note, |derived_esk| {
+            if D::epk_bytes(&D::ka_derive_public(&note, &derived_esk))
+                .ct_eq(&epk_bytes)
+                .into()
+            {
+                NoteValidity::Valid
+            } else {
+                NoteValidity::Invalid
+            }
+        })
+    }
+}
+
+/// Trial decryption of the compact note plaintext by the recipient for light clients.
+///
+/// Attempts to decrypt and validate the first 52 bytes of `enc_ciphertext` using the
+/// given `ivk`. If successful, the corresponding Sapling note is returned, along with the
+/// `PaymentAddress` to which the note was sent.
+///
+/// Implements the procedure specified in [`ZIP 307`].
+///
+/// [`ZIP 307`]: https://zips.z.cash/zip-0307
+pub fn try_compact_note_decryption<D: Domain, Output: ShieldedOutput<D>>(
+    domain: &D,
+    ivk: &D::IncomingViewingKey,
+    output: &Output,
+) -> Option<(D::Note, D::Recipient)> {
+    assert_eq!(output.enc_ciphertext().len(), COMPACT_NOTE_SIZE);
+
```

Comment:
To match the spec, I would prefer that we change D::kdf to take ephemeral_key as a byte array. Since it would be too disruptive to change the code using this API to pass in a byte array, we can instead call `epk_bytes` here, and add a comment explaining why it is ok to diverge from the spec in that way, i.e. that the check for small order points excludes all non-canonical encodings.